src/qdlutils/applications/qdlscope/main.py

```diff
@@ -0,0 +1,431 @@
+        for sample in self.application_controller.read_counts_continuous(
+                            sample_time = self.daq_parameters['sample_time'], 
+                            get_rate = self.daq_parameters['get_rate']):
+
+            # Logging the measurement time directly for each sample
+            # This is probably inefficient and will decrease the actual sample rate
+            # Although it should be negligible.
+            # In this current configuration the logged time corresponds to the end of the
+            # sample time bin.
+            self.data_x.append(time.time() - start_time)
+            # Save the data
+            self.data_y.append(sample)
+            # Update the viewport
+            self.view.update_figure()
+
+            # If the length of the list is too long then terminate the experiemnt
+            if len(self.data_x) > self.max_allowed_samples:
+                logger.warning(f'Maximum number of allowed samples ({self.max_allowed_samples}) reached, stopping sampling.')
+                self.stop_sampling()
+
+
+        # Increment the total measurement time
+        self.total_measurement_time += self.application_controller.readout_time
+
+        logger.info('Sampling complete.')
+        try:
+            pass
+        except Exception as e:
+            logger.info(e)
+
+    def stop_sampling(self, tkinter_event: tk.Event = None) -> None:
+        '''
+        Callback function to stop the sampling.
+
+        Parameters
+        ----------
+        tkinter_event : tk.Event
+            The `tkinter` event (not used).
+        '''
+        # Catch if already stopped
+        if not self.application_controller.running:
+            return None
+
+        logger.info('Stopping sampling.')
+
+        # Set the running flag to false to stop after the next sample
+        self.application_controller.running = False
+
+        # Enable the buttons
+        self.view.control_panel.start_button.config(state='normal')
+        self.view.control_panel.pause_button.config(state='disabled')
+        self.view.control_panel.reset_button.config(state='normal')
+        self.view.control_panel.save_button.config(state='normal')
+
+    def reset_data(self, tkinter_event: tk.Event = None) -> None:
+        '''
+        Callback function to reset the scanner, resetting the data and figure.
+
+        Parameters
+        ----------
+        tkinter_event : tk.Event
+            The `tkinter` event (not used).
+        '''
+        # Catch if already running
+        if self.application_controller.running:
+            return None
+
+        logger.info('Resetting data.')
+
+        # Reset the data variables
+        self.data_x = []
+        self.data_y = []
+        self.total_measurement_time = 0
+
+        # Reset the figure
+        self.view.initialize_figure()
+
+        # Enable the buttons
+        self.enable_buttons()
+
+    def save_data(self, tkinter_event=None) -> None:
```

Comment:
saving on scope doesn't work:

Exception in Tkinter callback
Traceback (most recent call last):
  File "C:\Users\SiV Microscope\anaconda3\envs\test-qd\Lib\tkinter\__init__.py", line 1968, in __call__
    return self.func(*args)
           ^^^^^^^^^^^^^^^^
  File "C:\Users\SiV Microscope\Desktop\Scan Code\qdl-utils\src\qdlutils\applications\qdlscope\main.py", line 287, in save_data
    self.parent_application.last_save_directory = file_path # Save the last used file path
    ^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'ScopeApplication' object has no attribute 'parent_application'